### DISCUSSION TITLE
DeZog as a reverse engineering tool

### DISCUSSION THREAD

**Discussion by lmartorella** [maziac/DeZog#83] [Ideas]
The disassembler view of DeZog, with his support to debug/source files, is awesome for brand new code, but it is even more intriguing when stepping into old assembly code dumped from some ROM or dusty tape.

In that case, it would be fantastic if the map/listing file could be edited in-place during debugging. I would be interested in:
- Give meaningful names to automatic generated labels (`Lxxxx`)
- Add comments
- Introduce constants

The editing should happen from the main `disasm` view. Everything should be saved down during editing in order to be saved and versioned. 

Since the "source" will likely be a binary file, so not bound to a specific assembler/compiler, the most convenient format could be used (lis? map?).

What do you think?

Thanks, L

**Comment by maziac** [comment]
Have you taken a look at [z80dismblr](https://github.com/maziac/z80dismblr)?
It is a standalone disassembler (I'm using it also inside DeZog but it is also a standalone project).
When using standalone you have a few more options.

Please have a look at the ["Interactive" Usage](https://github.com/maziac/z80dismblr#interactive-usage).

I haven't worked with this for a longer time but I think this is similar to what you are after.

A workflow would look like:
1. run z80dismblr
2. start DeZog with the disassembly
3. find more labels / meanings of labels
  a) add the address/labels to a file
  b) you can additionally also add comments !
4. run z80dismblr with that address/labels file
5. Goto 2

**Reply by lmartorella** [reply]
Great, that z80dismblr project is really what one can expect from a disassembler tool.
However, during reverse engineering via a simulator/debugger, it often happen that comments and labels should be assigned during the live debugging sessions (e.g. after have simulated a specific input or machine state).

The Visual Studio Code IDE is quite powerful in terms of programming paradigms, and "refactoring" tools are becoming widespread in all the supported language. Now, I am not absolutely asking to implement the "Rename Symbol" in the Z80 assembler (it would be something like 40 years late :).
However I would like to see the above steps 3/4 automated in such way: editing the comments/addresses in the list file can automatically update the disassembler view without losing the machine state of the current debugging session?
Obviously, the next step would be allowing editing the comments and the labels directly from the disassembly view (and the list file to be automatically keep up-to-date).

Thanks! L

**Comment by maziac** [comment]
Thinking about using DeZog for reverse engineering I have this idea to discuss:

DeZog can (already) associate list files with PC addresses.

At the moment list files are only read at the start of a debug session.
To make this useful for the reverse engineer the process could be made interactive.

E.g. a session could look like:
1. Start debug session (completely without list file, only object code).
2. DeZog disassembles part of the code.
3. The reverse engineer analyzes, steps through the code.
4. The reverse engineer understood a subroutine and copies the code from the disassembly file to the list file.
As this is in list file format every line starts with an address.
5. The reverse engineer adds comments and changes labels.
6. The reverse engineer **presses a button to re-read the list file** without loosing machine state.
7. DeZog throws away any former labels and address/file associations and re-reads this information from the list file.
8. DeZog's disassembly will not show the part of the list file anymore as this now has an address/file association.
9. The reverse engineer will now go on to the next subroutine.
10. Goto 3

I.e only by one the reverse engineer adds more and more to the list file until it completely covers all code.

**Comment by maziac** [comment]
Here is a first implementation for everyone interested to try out.
 It's on the develop branch. You have to build yourself (no release yet).
At the moment it's kinda working (tag rev_eng_1).
The process is described here: 
https://github.com/maziac/DeZog/blob/develop/documentation/ReverseEngineeringUsage.md

**Reply by lmartorella** [reply]
Will try, thanks!

**Comment by maziac** [comment]
The example is for mame. But this works similar in zsim.

**Comment by maziac** [comment]
Currently i'm refactoring a lot. The most recent development branch is not compilable.
I.e. you should stick to the tag rev_eng_1 for now.

**Comment by maziac** [comment]
DeZog 3.1.0 is out now, please have a look.
